### ISSUE TITLE
Create a new interface so that we can instanciate different kind of repositories

### ISSUE THREAD

**Issue by qparis** [PhoenicisOrg/phoenicis#1723]

**Comment by plata** [comment]
My understanding:
A repository contains a script which can dynamically populate the repository or a new repository.
Example: The Wine script populates the repository with all available Wine versions.
With this approach, the apps tab implementation could be reused for the engines tab.

Other possible use cases:
- Steam repository with naive implementations for all Steam apps
- same for Uplay, GoG...
- automatic Lutris repository

**Comment by qparis** [comment]
So you are introducing dependencies between repositories

**Comment by plata** [comment]
Or we have a new `RepositoryGenerator` type which is a single Javascript that can be like a repository.

**Comment by qparis** [comment]
Or inside a repository, we can add a optional set of class that will "decorate" the way the repository is loaded. Then, we create a separate git repository for all the new type of scripts

**Comment by plata** [comment]
The possible issue I see with this is the following: Where to put the generated scripts in the repository structure? You cannot allow to put it in the existing structure because it might overwrite existing scripts. If you use a separate repository which is completely generated, you solve this issue.

**Comment by qparis** [comment]
We add " - Plugins" directory. So it would be like 
- Applications 
- Engines 
- Utils
- Plugins

Inside a plugin, you'll find two interface implemntations (for now) like: 
- RepositoryListPopulationHook {
    List<TypeDTO> doFilter(List<TypeDTO> input);
}

- ScriptFactoryHook {
    ApplicationDTO hookApplication(ApplicationDTO input);

    String hookScript(String input); 
} 

Then repository can extends has many interface like this they want

**Comment by plata** [comment]
I don't quite see how this would behave yet. Can you give an example?

**Comment by qparis** [comment]
When the backend loads a repositorys, it does two things AFAIK:

- Its get a list of TypeDTO: The first interfaces allow you to change this list of TypeDTO to another one. They are chained so you can add as much as you need. 

- It gives assistance to run a script: The second interface can: 
   - change the content of the ApplicationDTO (miniature, name, everything)
   - decorate the script with a decorator (let's call it foo). When a script is decorated with a decorator foo, then foo(script) will be executed instead of script

**Comment by plata** [comment]
How does this add categories/applications to the repository?

**Comment by qparis** [comment]
By hooking TypeDTO (that contains Categories and games)

**Comment by plata** [comment]
Ok. If you have multiple repositories, the hook will apply to all previously loaded?

**Comment by qparis** [comment]
No, just the current one

**Comment by plata** [comment]
That should work. And it allows us to e.g. use naive Steam scripts only for cases where we do not have another script already.

I don't see the use cases for the decorators yet.

**Comment by qparis** [comment]
Let's imagine that you have a JSON script in your repository (before or after the hook is applied, it is not important). 

Then, the backend will execute decorator1(decorator2(decorator3(decorator4(json_script))), instead of json_script

**Comment by plata** [comment]
I understand how it works but not what it could be used for.

**Comment by qparis** [comment]
> That should work. And it allows us to e.g. use naive Steam scripts only for cases where we do not have another script already.
However, the hooks only have the current repository as input to avoid side effect

> I understand how it works but not what it could be used for.

Running JSON script without having an interpreter

**Comment by plata** [comment]
> Running JSON script without having an interpreter

So you would store Lutris scripts in the repository directly and then use the decorator at runtime? My idea was to create a Phoenicis format in the population hook already.

**Comment by qparis** [comment]
> > Running JSON script without having an interpreter
> 
> So you would store Lutris scripts in the repository directly and then use the decorator at runtime? My idea was to create a Phoenicis format in the population hook already.

One class, one responsibility. In addition, you'll have a lot of duplicated code if you do like this. And also, this will allow to support home made YAML/JSON/any kind of format script

**Comment by plata** [comment]
With your approach, the Lutris yaml must be downloaded inside a script.js, correct?

**Comment by qparis** [comment]
I don't think so because ScriptDTO contains a URI to the script content. So I'm not sure

**Comment by plata** [comment]
In this case it will never work offline.

**Comment by qparis** [comment]
Phoenicis automatically caches the URLs, right?

**Comment by plata** [comment]
It caches the git repository. If the hook does not create files but only DTO objects in the repository tree, they will not be there.

**Comment by qparis** [comment]
So how does it work between the String source and the URL source?

**Comment by plata** [comment]
It doesn't really change it. Then you will just write it to that file in the cache. Additionally, I think it breaks the concept because it avoids that you can always run the scripts in a repository. Let's say you have two repositories with Lutris scripts. Then you will always need the decorators in both repositories.

**Comment by qparis** [comment]
Yes. But introducing dependencies between repositories is very hazardous. Imagine someone creates a hook that breaks an official script and we get report about this

**Comment by qparis** [comment]
The caching issue should be treated globally. If we have a HTTP repository with HTTP urls for scripts, the caching needs to be supported. If I understand, we are currently relying on local git to support the cache, which is I think a problem

**Comment by plata** [comment]
Regardless how we realize the caching: I don't see the advantage of caching the yaml instead of a script based on the yaml which would have the advantage that you can run it directly.

**Comment by qparis** [comment]
I feel uncomfortable with he idea of transpiling source code. It smells hacky

**Comment by plata** [comment]
We don't have to do that. I was thinking more of something like you did in the POC PR:
```Javascript
include(["engines", "wine", "quick_script", "lutris_wine_steam_script"]);

var installerImplementation = {
    run: function () {
        new LutrisWineSteamScript()
            .yaml("yaml content here")
            .go();
    }
};

/* exported Installer */
var Installer = Java.extend(org.phoenicis.scripts.Installer, installerImplementation);
```

**Comment by qparis** [comment]
So you are kind of transpiling, because you are generating some JS Files. 
With my solution, no JS file is generated at all

**Comment by plata** [comment]
I know but in my opinion that's rather a problem because you save a yaml in a js and you cannot execute some of the scripts (more or less randomly).

**Comment by qparis** [comment]
You can because you have the hook installed. 

I mean, if you generate JS files, your method is very tied to the language of the scripts. Your generated code will be stored inside scripts, so codacy will not apply, the refactor function will not work (Imagine you rename some of the imports you are using in the generated script, ...) 

Etc..

Generation of code should be generally avoided I think

**Comment by qparis** [comment]
Not to mention security problem. What if we have a .yaml file that contains crafted JS injection

**Comment by plata** [comment]
How would it work for the Wine engine versions?

**Comment by qparis** [comment]
What do you mean?

**Comment by plata** [comment]
Remember that the idea was also to use this mechanism for Wine versions such that we could reuse the apps tab implementation?

**Comment by qparis** [comment]
I don't remember that

**Comment by plata** [comment]
I don't know yet. I'm not really convinced by any of the solutions.

**Comment by qparis** [comment]
I dont see what could be the problem if we decide to store the scripts in another format.

**Comment by plata** [comment]
It's not the format (we could rename it from .js and I would be fine with that).

I think my issue is somewhere else: With this approach, the meaning of script.js is not well defined anymore. Currently, it is a script which is triggered from the embedding application but then is in charge of the execution control. If the script is only e.g. a yaml, it's meaning changes logically because it's suddenly no longer executing but **is** executed (by the decorator). Mixing these two approaches doesn't feel good for me.

Maybe we should just avoid to cache these external, dynamically created parts of the repository. Then we could simply use the populate hook to create executable scripts from yaml like I've described in https://github.com/PhoenicisOrg/phoenicis/issues/1723#issuecomment-451653839. Redundancy would not be a problem because all the logic is inside the hook. Additionally, it would allow us to create installable engine version scripts which could be used in the same way as apps.

**Comment by qparis** [comment]
> It's not the format (we could rename it from .js and I would be fine with that).
> 
> I think my issue is somewhere else: With this approach, the meaning of script.js is not well defined anymore. Currently, it is a script which is triggered from the embedding application but then is in charge of the execution control. If the script is only e.g. a yaml, it's meaning changes logically because it's suddenly no longer executing but **is** executed (by the decorator). Mixing these two approaches doesn't feel good for me.

The best approach is to consider that a script is executed. This is a good IoC principle we should apply. By the way, this is what you have started to do by enforcing script to extend objects 


> 
> Maybe we should just avoid to cache these external, dynamically created parts of the repository. Then we could simply use the populate hook to create executable scripts from yaml like I've described in [#1723 

This approach is very dangerous.

**Comment by plata** [comment]
I would like to hear @madoar opinion as well.

**Comment by madoar** [comment]
@qparis I don't understand what you would like to populate a repository with?

@plata My idea of Phoenicis is, that it should stay in control at all times.
Ideally this means, that the every command executed inside a "script" should be checked/performed by Phoenicis.

The reason why I think this is, that this approach leads to more security and control.
Our scripts are so generic, that you can do anything with them.
This in turn makes it very difficult to find bugs later on.
In addition it makes it quite easy for script developers to inject some malicious commands into their scripts.

In addition I think Phoenicis should only support one type/format of script officially.
Everything else should be done by the community externally from Phoenicis.
Otherwise it can become a nightmare to maintain this.
This means that if there are people who want to have direct/native support of lutris scripts in Phoenicis, they should write a solution **outside** of Phoenicis.
This is also the reason for my idea of a converter for lutris scripts, because such a converter could exist outside of Phoenicis.

**Comment by qparis** [comment]
> @qparis I don't understand what you would like to populate a repository with?
> 

Almost nothing apart from:
* A quick Script
* a plugin 

> @plata My idea of Phoenicis is, that it should stay in control at all times.
> Ideally this means, that the every command executed inside a "script" should be checked/performed by Phoenicis.
> 

This is ideally but unfortunately, when we delegate our script to any program (wine, steam, gog, ....) this is no longer true

> The reason why I think this is, that this approach leads to more security and control.
> Our scripts are so generic, that you can do anything with them.
> This in turn makes it very difficult to find bugs later on.
> In addition it makes it quite easy for script developers to inject some malicious commands into their scripts.

This is true only if we try to generate source files. If we evaluate the script commands at runtime, we have full control on the execution flow. Except for the ‘exec’ commands that we can decide to not support on a first stage.

In addition, if you use a converter like you suggest, you may end up with thousands generated scripts that will never be checked one by one, so the situation is almost the same.

> 
> In addition I think Phoenicis should only support one type/format of script officially.
> Everything else should be done by the community externally from Phoenicis.
> Otherwise it can become a nightmare to maintain this.

This is why I suggest to enforce the creation of a new repository to prevent plugins from modifying and official repository. It makes things clear separate and we do not force user support to subscribe to these new repositories.

> This means that if there are people who want to have direct/native support of lutris scripts in Phoenicis, they should write a solution **outside** of Phoenicis.
> This is also the reason for my idea of a converter for lutris scripts, because such a converter could exist outside of Phoenicis.

I think that trying to generate source files almost always lead to problems.
It is very hard to implement the right parsers, it can lead to security problems, and it requires a lot of effort to maintain. On the other hand, a code that evaluates a JSON script can be made clear to read (this is what Lutris itself or wine do after all)

**Comment by qparis** [comment]
A fourth approach would be to allow repositories to store resource files

**Comment by plata** [comment]
We have .reg files for some scripts?

**Comment by qparis** [comment]
They are included in the script source file, don’t they?

**Comment by qparis** [comment]
I think we should not treat this problem as if we wanted to support other script languages. This is not the case. 
In fact JSON/YAML are not script languages. They are not turing-complete, they do not contain any complex stuff like conditions, loops, etc.. They just contain a description of what needs to be done. 

Therefore, they can be regarded as a data input (like a URL, or a game ID) and not as a script that needs to be executed. This makes a lot of differences in the approach. This is also why I think we can treat them as resource.

So in a nutshell, I think we have two things we need to treat separately: 
1. Metadata: How a repository can read a remote source of script, and expose them as a phoenicis native source of script.
2. Script content: How can we automatically inject external input (URL or game ID for GOG, JSON for Lutris, etc...) inside a pre-written JS script (like the ones I have made in my previous requests).

Don't know if I'm clear or not

**Comment by plata** [comment]
https://github.com/PhoenicisOrg/scripts/tree/4672e1c56b79edf6e9280aaaa3cf350ad1863e3a/Applications/Games/Subnautica/resources

We can use this to store other resources as well (e.g. yaml).

**Comment by qparis** [comment]
Cool, so we already have that

**Comment by plata** [comment]
Yep. The question is how we use it. Simplest approach would be to have 1 Lutris script and all yaml in the resources. Then the script shows a list in the wizard. We could put this in a new category "special".

**Comment by qparis** [comment]
I don't like that. In term of user experience, it does not have any sense

**Comment by plata** [comment]
How would you want to handle it?

**Comment by qparis** [comment]
It makes more sense to concatenate the games with the ones of our repository

**Comment by plata** [comment]
Yes but technically?

**Comment by qparis** [comment]
They won't be stored I supposed, the TreeDTO will just be hooked

**Comment by plata** [comment]
Even in that case, they must be located somewhere.

**Comment by qparis** [comment]
Directly from the remote servers

**Comment by qparis** [comment]
At the moment, we are relying on jgit to manage to script caching. We should not do that. 
I think the script content should not be stored in the DTO tree by default. A component that manages the cache should do it. This component would get the script from the URL and store it inside the cache

**Comment by madoar** [comment]
@qparis what do you mean by `relying on jgit to manage to script caching`? I think we will should change our `Repository` interface to contain three methods, which should be independent in their execution:

```java
List<Engine> getEngines(...);
List<CategoryDTO> getCategories(...);
ScriptDTO getScript(List<String> path);
```
When looking at our current `Repository` interface I'm also not too happy with the `List<String> path` parameter we pass to nearly every method, especially because it's not really a path but more a concatenation of ids, which require a lot of computation before the (id-)"path" can be matched. More natural in my opinion would be for example the following method signature:

```java
ScriptDTO getScript(ApplicationDTO application, String scriptId);
```

**Comment by qparis** [comment]
If I understand well, we consider that the script are available locally because they are stored in a local git repository right? 

When is the content of the script is populated inside the DTO?

**Comment by madoar** [comment]
This basically happens in `LocalRepository#fetchScripts`, which is called in a nested/recursive fashion. Here the call tree:

1. `LocalRepository#fetchInstallableApplications`
2. `LocalRepository#fetchTypes`
3. `LocalRepository#fetchCategories`
4. `LocalRepository#fetchApplications`
5. `LocalRepository#fetchScripts`

**Comment by qparis** [comment]
I propose to do it when the script is run

**Comment by madoar** [comment]
The question is then: how would you do the script lookup?
To fetch all scripts belonging to an application, the following information is required:
```java
fetchScripts(String typeId, String categoryId, String applicationId, File applicationDirectory)
```
To fetch only a single script, you additionally need the `scriptId`.
I'm not sure you have this information during script execution.

**Comment by qparis** [comment]
You must have it because JavaFX has it right?

**Comment by madoar** [comment]
No, we require these parameters because of the way our file system repositories are structured and because of how we access entities via our `List<String> path` objects.

### Let's give an example
Let us assume we want to fetch the script located in `~/Applications/Development/Notepad++/v7.2.2/script.js`.
Each script has a unique "path", which is used to locate the script. For Notepad++ the path is `["applications", "development", "notepad_plus_plus", "v_7_2_2"]`. Please note, that the used path differs from the filesystem path! (which is the reason why the scripts are stored in the `ScriptDTO` class)

When fetching the script for the above mentioned path `["applications", "development", "notepad_plus_plus", "v_7_2_2"]` the algorithm uses the following approach:

1. Search for the folder in `~/` which contains a `type.json` file with the field `id: "applications"`. This folder is called `type-folder` in the following
2. Search in the found `type-folder` for a subfolder which contains a `categories.json` file with the field `id: "development"`. This folder is called `category-folder` in the following
3. Search in the found `category-folder` for a subfolder which contains an `application.json` file with the field `id: "notepad_plus_plus"`. This folder is called `application-folder` in the following
4. Search in the found `application-folder` for a subfolder which contains a `script.json` file with the field `id: "v_7_2_2"`. This folder is called the `script-folder` in the following
5. Return the `script.js` file located in the `script-folder`

The problem is this approach is, that we need to do the `json` file lookups to match the paths.

I think to solve this problem we would need a different approach to lookup scripts.

**Comment by plata** [comment]
And to add to this: The resources we were talking about have to be part of the repository DTO tree as well such that they can be used (at least as far as I'm aware).

**Comment by qparis** [comment]
I understand these steps. 

I think we should change this so that: 
- When POL is loaded, the TreeDTO without the script content is fully populated (whatever is the implementation you chose). 
- When you run a script, the URL is read and the script is fetched

Therefore, you can access the URL from the script path, right? 

Same for the resources.

**Comment by plata** [comment]
If we do it like this, you can only install Lutris if you are online. The approach cannot work for resources because there is no such thing like a URL for resources.

**Comment by qparis** [comment]
There is, everything is a URL. 
We can then implement a cache mechanism on top of that that works for every type of repository

**Comment by qparis** [comment]
Better: 
- By default: no caching (better loading performance)
- If the user wants, he/she can enable an "offline mode" that will automatically cache all the scripts when the repository is updated. 

The populate will add scripts to the DTO tree dynamically

**Comment by qparis** [comment]
Even better: Implement several caching strategies for resources: 
 - No cache
 - Always up to date
 - Cached for a long time (miniatures ?)

**Comment by plata** [comment]
I think we should separate this issue from caching.

**Comment by qparis** [comment]
Yes

**Comment by plata** [comment]
So:

> The populate will add scripts to the DTO tree dynamically

I thought you wanted to use resources instead of scripts?

**Comment by qparis** [comment]
Script in a large sense i.e. Script URL (note that several scripts can use the same URL), resources, metadata

**Comment by qparis** [comment]
The only thing is that the plugin should not be stored inside URL because they need to be always available

**Comment by plata** [comment]
Yes, it can create new applications. Does this mean now, that you want to create an application per Lutris script. Because that was my understanding initially but you didn't seem to like it (redundant code etc.).

**Comment by qparis** [comment]
No, the plugin will do it for you

**Comment by plata** [comment]
Sure, the plugin will create the application.

In this case, I think we are on the same page. The population hook creates an application per Lutris script (similar to the one you have in the scripts repo) which can then be executed like any other script.

**Comment by madoar** [comment]
@qparis I don't understand the difference between your `cache` idea and a `WebRepository` implementation?

**Comment by qparis** [comment]
WebRepository is a particular implementation that would take advantage (or not) of the cache system. The cache system I imagine is agnostic from the kind of repository

**Comment by madoar** [comment]
So the cache system is basically a wrapper `Repository` implementation, which takes another repository and caches its content?

**Comment by madoar** [comment]
This is still hard for me to imagine what exactly you're planning.
In my opinion every change to the `Repository` implementations and the associated manager classes, that decouple the fetch methods is a good change.

I think I need a basic prototype PR to really understand what you want to do.

**Comment by plata** [comment]
Agree to need of a prototype.
